marimo/_ast/parse.py

```diff
@@ -0,0 +1,850 @@
+                    end_lineno = node.body[-1].lineno
+                    end_col_offset = node.body[-1].col_offset
+
+            # We're in the case where we have something like
+            # @app.cell
+            # def foo():
+            #   # Just comments
+            #   return
+            else:
+                # If we are on the same line as the return statement,
+                # just return a blank cell.
+                if node.lineno + lineno - node.body[0].lineno == 0:
+                    return CellDef(
+                        code="",
+                        options=kwargs,
+                        lineno=node.lineno + lineno,
+                        col_offset=node.col_offset + col_offset,
+                        end_lineno=node.lineno + lineno,
+                        end_col_offset=len(self.lines[-1]),
+                        name=getattr(node, "name", DEFAULT_CELL_NAME),
+                    )
+                else:
+                    end_lineno = node.body[-1].lineno - 1
+                    end_col_offset = None
+        cell_code = self.extract_from_offsets(
+            lineno=node.lineno + lineno - 1,
+            col_offset=col_offset,
+            end_lineno=end_lineno - 1,
+            end_col_offset=end_col_offset,
+        )
+
+        # NB: Feels pretty hacky, would love if someone had a suggestion for this
+        # But the only other better way (I can see) is to drop down to the
+        # toeknizer lever and handle it there + handle particular edge
+        # cases.
+        #
+        # If the last value is not a return statement, we need to grab
+        # trailing comments.
+        if not has_return and not single_line:
+            # Determine leading spaces
+            leading_spaces = len(cell_code) - len(cell_code.lstrip())
+            indent = cell_code[:leading_spaces]
+
+            # Attempt to keep adding lines, and ensure it dedents to the correct
+            # level
+            new_end = end_lineno - 1
+            for new_end in range(end_lineno, len(self.lines)):
+                if not self.lines[new_end].strip() or self.lines[
+                    new_end
+                ].startswith(indent):
+                    end_col_offset = len(self.lines[new_end])
+                    continue
+                new_end -= 1
+                break
+            if new_end > end_lineno:
+                cell_code += "\n".join(
+                    self.lines[end_lineno : new_end + 1]
+                ).rstrip()
+                end_lineno = new_end - 1
+
+        if end_col_offset is None:
+            end_col_offset = 0
+        return CellDef(
+            code=dedent(cell_code),
+            options=kwargs,
+            lineno=node.lineno + lineno - 1,
+            col_offset=node.col_offset + col_offset,
+            end_lineno=_none_to_0(node.end_lineno) + end_lineno,
+            end_col_offset=_none_to_0(node.end_col_offset) + end_col_offset,
```

Comment:
Can you explain the various offsets to `lineno` and `col_offset` here in a comment?